pkg/executor/sortexec/sort.go

```diff
@@ -119,57 +120,172 @@ func (e *SortExec) Next(ctx context.Context, req *chunk.Chunk) error {
 		e.fetched = true
 	}
 
-	if len(e.PartitionList) == 0 {
+	sortPartitionListLen := len(e.sortPartitions)
+	if sortPartitionListLen == 0 {
 		return nil
 	}
-	if len(e.PartitionList) > 1 {
-		if err := e.externalSorting(req); err != nil {
+
+	if sortPartitionListLen == 1 {
+		if err := e.onePartitionSorting(req); err != nil {
 			return err
 		}
 	} else {
-		for !req.IsFull() && e.Idx < e.PartitionList[0].NumRow() {
-			_, _, err := e.PartitionList[0].GetSortedRowAndAlwaysAppendToChunk(e.Idx, req)
-			if err != nil {
-				return err
-			}
-			e.Idx++
+		if err := e.externalSorting(req); err != nil {
+			return err
 		}
 	}
 	return nil
 }
 
+func (e *SortExec) initExternalSorting() error {
+	e.multiWayMerge = &multiWayMerge{e.lessRow, make([]rowWithPartition, 0, len(e.sortPartitions))}
+	for i := 0; i < len(e.sortPartitions); i++ {
+		// We should always get row here
+		row, err := e.sortPartitions[i].getNextSortedRow()
+		if err != nil {
+			return err
+		}
+
+		e.multiWayMerge.elements = append(e.multiWayMerge.elements, rowWithPartition{row: row, partitionID: i})
+	}
+	heap.Init(e.multiWayMerge)
+	return nil
+}
+
+func (e *SortExec) onePartitionSorting(req *chunk.Chunk) (err error) {
+	err = e.sortPartitions[0].checkError()
```

Comment:
If the check pass and then the partition trigger spilling, what happened? I think we should check error in getNextSortedRow, not SortExec's function.

Reply:
> If the check pass and then the partition trigger spilling, what happened? I think we should check error in getNextSortedRow, not SortExec's function.

Next call for `onePartitionSorting` will catch this error, I think it's enough.